### PR TITLE
copy editing constraint pages

### LINKED ISSUE #313: sql: Copyedit Constraints
Copyedit/formatting: https://www.cockroachlabs.com/docs/data-definition.html#constraints

### PR DESCRIPTION
**Pull request by sploiselle**
- Edited pages for constraints
- Unified style for referencing constraints using format *<Constraint Name in all caps> constraint*. For example, "the Primary Key constraint". When talking about a concept and not a constraint itself, use the term without any caps. For example, "primary key".
- Changed references of "Default value" to "Default Value."

Closes #313

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/964)

### CODE REVIEW

**Comment by dt** [comment]
Not sure if we care, but lots of cases where we say things restrict values *inserted* but it constraints all writes, including UPDATES.

Mentioned in one place but applicable elsewhere: no difference between expressing a constraint on a col or at table level except for NULL and DEFAULT. We should probably document the table level and then only mention the single-col as a shorthand.

---

Reviewed 4 of 25 files at r1.
Review status: 4 of 25 files reviewed at latest revision, 11 unresolved discussions.

---

*[check.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDMJrLvdosRaI-V0JL:-KaDMJrLvdosRaI-V0JM:buxjpgg) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/check.md#L13)):*
> ```Markdown
> ## Details
> 
> - If you add a Check constraint to an existing table, existing values are not checked. However, any updates to those values will be.
> ```

We might want to note that this is is a temporary limitation, not a permanent design decision -- in the future, the modifier `NOT VALID` would trigger this behavior (where the check is added quickly but only applies to new writes), but the default will be to check existing data. Currently we always do the `NOT VALID` behavior only because we haven't implemented queuing up the validation.

---

*[check.md, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDMxPUKv2Q84xCaTtn:-KaDMxPUKv2Q84xCaTto:b-ukt59k) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/check.md#L103)):*
> ```Markdown
> 
> - [Constraints](constraints.html)
> - [`DROP CONSTRAINTS`](drop-constraint.html)
> ```

Should this be plural?

---

*[constraints.md, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDN3mXOYJEcfOsHA_X:-KaDN3mXOYJEcfOsHA_Y:b-25od0q) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/constraints.md#L7)):*
> ```Markdown
> ---
> 
> Constraints provide additional data integrity beyond SQL's type checking. They let you define additional parameters for a column's data, which are checked whenever values are manipulated (inserted, deleted, or updated) and reject modifications that violate the constraint.
> ```

I don't know if I'd really think about them as related to type checking -- it's more about domain/business logic specific to your application being additionally enforced by the DB -- you _could_ (and probably _should_, to have more control over error handling) check these things in your application, but also adding constraints on the db itself is an additional layer of safety, particularly for something like FKs where the app may operate on related tables separately.

---

*[constraints.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDOEqLUG6cGpIqhbKU:-KaDOEqLUG6cGpIqhbKV:b-omqn3d) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/constraints.md#L9)):*
> ```Markdown
> Constraints provide additional data integrity beyond SQL's type checking. They let you define additional parameters for a column's data, which are checked whenever values are manipulated (inserted, deleted, or updated) and reject modifications that violate the constraint.
> 
> For example, the Unique constraint requires that all values in a column be unique from one another (except *NULL* values). If you attempt to insert a duplicate value, the constraint rejects the entire statement.
> ```

updates can also create dupes.

---

*[not-null.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDa-gXDxz5jUUJ6ey7:-KaDa-gYnR1EVnjAz5Ln:bk4coxl) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/not-null.md#L23)):*
> ```Markdown
>   CockroachDB tries to write a *NULL* value into column `b`. If that column has the Not Null constraint, the `INSERT` statement is rejected.
> 
> - You can only define the Not Null constraint when [creating a table](#syntax); you cannot add it to an existing table. However, you can [migrate data](constraints.html#table-migrations-to-add-or-change-immutable-constraints) from your current table to a new table with the constraint you want to use.
> ```

might want to note this is only current limitation, not a long-term decision.

---

*[not-null.md, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDdSBdiRNUP7XRMQZv:-KaDdSBdiRNUP7XRMQZw:broctjq) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/not-null.md#L75)):*
> ```Markdown
> 
> - [Constraints](constraints.html)
> - [`DROP CONSTRAINTS`](drop-constraint.html)
> ```

plural?

---

*[unique.md, line 3 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDZDq0GQW-qwXZ7OL-:-KaDZDq0GQW-qwXZ7OL0:b-ck4hzy) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L3)):*
> ```Markdown
> ---
> title: Unique Constraint
> summary: The Unique constraint specifies that each value in the constrained column must be unique (though *NULL* values are still allowed).
> ```

"still" seems strange to me (ditto below).
maybe just "non-null values must be unique?

---

*[unique.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDRRu4xSr97U1KiXwi:-KaDRRu4xSr97U1KiXwj:b-7oho6) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L13)):*
> ```Markdown
> ## Details
> 
> - You can still insert *NULL* values into columns with the Unique constraint because *NULL*s are not considered equal to other *NULL*s and therefore don't appear as the same value. This means that it's possible to insert rows that appear to be duplicates if one of the values is *NULL*.
> ```

I might lead with the fact that NULL is the absence of a value, so it is never equal and thus not considered a duplicate value. Thus if one uses NULL for one of the columns in a constraint etc.

---

*[unique.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDc3Vv0KBJyFLOb1MI:-KaDc3Vv0KBJyFLOb1MJ:b-8jdcqm) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L23)):*
> ```Markdown
> ## Syntax
> 
> The syntax of the Unique constraint depends on whether you want the constraint to use a [single column](#single-column-column-level) or [multiple columns](#multiple-column-table-level).
> ```

Eh, you can use the table-level constraint on a single col -- in fact, except for DEFAULT and NULL, the col-level one is just a shorthand that we translate to table level constraint (on  a single-col) before evaluating it, so that's what you'll see in `SHOW CREATE TABLE`. 

---

*[unique.md, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDPQUmh1C4tVu-nSU3:-KaDPQUmh1C4tVu-nSU4:b-o3zol1) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L73)):*
> ```Markdown
> ## Usage Example
> 
> Be aware that if a table has a Unique constraint on columns without the [Not Null constraint](not-null.html), it is possible to insert rows that appear to be duplicates if they contain a *NULL* value in at least one of the columns. This is because *NULL*s are never considered equal and hence don't violate the uniqueness constraint.
> ```

Maybe just me, but I found this one a little hard to follow / it sounded like we had some sort of defect that meant I needed to use NOT NULL for it to work, and it's duplicative with the statement above about nulls.

---

*[unique.md, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDPDXTHV2tn3-Eh3Nj:-KaDPDXTHV2tn3-Eh3Nk:broctjq) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L101)):*
> ```Markdown
> 
> - [Constraints](constraints.html)
> - [`DROP CONSTRAINTS`](drop-constraint.html)
> ```

plural?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/964#-:-KaDaItZymEQP3Q3q_OB:b-qxck3)*

**Comment by jseldess** [comment]
@sploiselle, would you mind incorporating this change into the foreign key doc as well? https://github.com/cockroachdb/docs/issues/990

Here's how Ben summarized it in this week's release notes: 
> The `ALTER TABLE VALIDATE CONSTRAINT` command can now be used to validate foreign keys in the `UNVALIDATED` state.

**Comment by sploiselle** [comment]
Review status: 4 of 25 files reviewed at latest revision, 11 unresolved discussions.

---

*[check.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDMJrLvdosRaI-V0JL:-KaEH27h_vTeirt_wf3U:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/check.md#L13)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We might want to note that this is is a temporary limitation, not a permanent design decision -- in the future, the modifier `NOT VALID` would trigger this behavior (where the check is added quickly but only applies to new writes), but the default will be to check existing data. Currently we always do the `NOT VALID` behavior only because we haven't implemented queuing up the validation.
</blockquote></details>

Done.

---

*[check.md, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDMxPUKv2Q84xCaTtn:-KaEHD31H7uFvFzx0ssJ:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/check.md#L103)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Should this be plural?
</blockquote></details>

Done.

---

*[constraints.md, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDN3mXOYJEcfOsHA_X:-KaEI5CPHZMnVA0T7qtw:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/constraints.md#L7)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I don't know if I'd really think about them as related to type checking -- it's more about domain/business logic specific to your application being additionally enforced by the DB -- you _could_ (and probably _should_, to have more control over error handling) check these things in your application, but also adding constraints on the db itself is an additional layer of safety, particularly for something like FKs where the app may operate on related tables separately.
</blockquote></details>

Done.

---

*[constraints.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDOEqLUG6cGpIqhbKU:-KaEI5uSufZ0RbxbyQtQ:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/constraints.md#L9)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

updates can also create dupes.
</blockquote></details>

Done.

---

*[not-null.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDa-gXDxz5jUUJ6ey7:-KaEIYYgVjUVKrhk6vOU:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/not-null.md#L23)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

might want to note this is only current limitation, not a long-term decision.
</blockquote></details>

Done.

---

*[not-null.md, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDdSBdiRNUP7XRMQZv:-KaEIZ-crm48WMMQacK_:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/not-null.md#L75)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

plural?
</blockquote></details>

Done.

---

*[unique.md, line 3 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDZDq0GQW-qwXZ7OL-:-KaEJDBIl6jxii-TmCZl:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L3)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

"still" seems strange to me (ditto below).
maybe just "non-null values must be unique?
</blockquote></details>

Done.

---

*[unique.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDRRu4xSr97U1KiXwi:-KaILYrvkH5WK5JzZoVs:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L13)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I might lead with the fact that NULL is the absence of a value, so it is never equal and thus not considered a duplicate value. Thus if one uses NULL for one of the columns in a constraint etc.
</blockquote></details>

Done.

---

*[unique.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDc3Vv0KBJyFLOb1MI:-KaIVZl9S0DkSESJL8-t:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L23)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Eh, you can use the table-level constraint on a single col -- in fact, except for DEFAULT and NULL, the col-level one is just a shorthand that we translate to table level constraint (on  a single-col) before evaluating it, so that's what you'll see in `SHOW CREATE TABLE`. 
</blockquote></details>

Done.

---

*[unique.md, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDPQUmh1C4tVu-nSU3:-KaIX49igp5s41lha4Dw:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L73)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Maybe just me, but I found this one a little hard to follow / it sounded like we had some sort of defect that meant I needed to use NOT NULL for it to work, and it's duplicative with the statement above about nulls.
</blockquote></details>

Done.

---

*[unique.md, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/964#-KaDPDXTHV2tn3-Eh3Nj:-KaIX4a0T-UzXDxKzXv_:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/2f6c637d27b16cfc44f1fa885c66566dc5ce990f/unique.md#L101)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

plural?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/964)*

**Comment by sploiselle** [comment]
@jseldess Just talked to @dt about `VALIDATE CONSTRAINT` and I'll open up a new PR with that stuff; has some substantial changes to the `ADD CONSTRAINT` doc.

---

Review status: 4 of 25 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/964#-:-KaJ-IXkLZaKrSU3kxpz:b7qlxuy)*

**Comment by dt** [comment]
:lgtm:

---

Review status: 3 of 25 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/964#-:-KaJct-egtReVm7j_EYW:bnfp4nl)*

File: not-null.md [hunk 1]
```diff
@@ -4,24 +4,39 @@ summary: The NOT NULL constraint specifies the column may not contain NULL value
...
 
 | Parameter | Description |
 |-----------|-------------|
 | `table_name` | The name of the table you're creating. |
 | `column_name` | The name of the constrained column. |
 | `column_type` | The constrained column's [data type](data-types.html). |
-| `column_constraints` | Any other column-level [constraints](constraints.html) you want to apply. |
+| `column_constraints` | Any other column-level [constraints](constraints.html) you want to apply to this column. |
 | `column_def` | Definitions for any other columns in the table. |
 | `table_constraints` | Any table-level [constraints](constraints.html) you want to apply. |
```
Comment (jseldess) [inline not-null.md]: Not sure if we need this example, since it's identical to the start of the **Usage Example** below.

File: primary-key.md [hunk 2]
```diff
@@ -4,21 +4,36 @@ summary: The Primary Key constraint specifies that the columns can be used to un
...
 
 <div id="toc"></div>
 
 ## Details
 
-- A Primary Key constraint requires that the column values are unique and **do not** contain *NULL* values. You can optionally give the constraint a name using the `CONSTRAINT name` syntax, otherwise the constraint and its associated unique index are called **primary**. 
-   {{site.data.alerts.callout_info}}Strictly speaking, a primary key's unique index is not created, it's derived from the keys under which the data is stored so it takes no additional space. However, it appears as a normal unique index when using commands like <code>SHOW INDEXES</code>.{{site.data.alerts.end}}
-- Columns that are part of a Primary Key are mandatory (NOT NULL). If an optional (nullable) column is made part of a Primary Key, it is made mandatory (NOT NULL). 
+- Tables can only have one primary key.
+- To ensure each row has a unique identifier, the Primary Key constraint combines the properties of both the [Unique](unique.html) and [Not Null](not-null.html) constraints. The properties of both constraints are necessary to make sure each row's primary key columns contain sets of values that do not appear to be the same.
```
Comment (jseldess) [inline primary-key.md]: nit: instead of "values that do not appear to be the same", can't we just say, "values that are not the same"?

File: unique.md [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
+- You can insert *NULL* values into columns with the Unique constraint because *NULL* is the absence of a value, so it is never equal to other *NULL* values and not considered a duplicate value. This means that it's possible to insert rows that appear to be duplicates if one of the values is *NULL*.
+  
+  If you need to strictly enforce uniqueness, use the [Not Null constraint](not-null.html) in addition to the Unique constraint. You can also achieve the same behavior through the table's [Primary Key](primary-key.html).
+
+- Columns with the Unique constraint automatically have an [index](indexes.html) created with the name `<table name>_<columns>_key`. To avoid having two identical indexes, you should not create indexes that exactly match the Unique constraint's columns and order. <br/><br/>The Unique constraint depends on the automatically created index, so dropping the index also drops the Unique constraint.
+- When using the Unique constraint on multiple columns, the collective values of the columns must be unique. This *does not* mean that each value in each column must Unique, as if you had applied the Unique constraint to each column individually.
```
Comment (jseldess) [inline unique.md]: `must Unique` --> `must be Unique`

**Comment by jseldess** [comment]
Oh, one additional request: in the sidenav, let's change `Foreign Keys` to `Foreign Key` to match the form of the other constraints.

**Comment by jseldess** [comment]
:lgtm:

---

Reviewed 17 of 25 files at r1, 1 of 7 files at r2, 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 30 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/964#-:-KbHMeujCMmACIYCRsxy:bnfp4nl)*
